src/signing/collector/signatures_collector.rs

```diff
@@ -0,0 +1,554 @@
+            .await
+            .inspect_err(|e| eprintln!("Failed to use factor sources: {:#?}", e));
+
+        self.outcome()
+    }
+}
+
+// === PRIVATE ===
+impl SignaturesCollector {
+    /// If all transactions already would fail, or if all transactions already are done, then
+    /// no point in continuing.
+    ///
+    /// `Ok(true)` means "continue", `Ok(false)` means "stop, we are done". `Err(_)` means "stop, we have failed".
+    fn continue_if_necessary(&self) -> Result<bool> {
+        self.state
+            .borrow()
+            .petitions
+            .borrow()
+            .continue_if_necessary()
+    }
+
+    async fn use_factor_sources(&self, factor_sources_of_kind: &FactorSourcesOfKind) -> Result<()> {
+        let interactor = self
+            .dependencies
+            .interactors
+            .interactor_for(factor_sources_of_kind.kind);
+        let factor_sources = factor_sources_of_kind.factor_sources();
+        match interactor {
+            // Parallel Interactor: Many Factor Sources at once
+            SigningInteractor::Parallel(interactor) => {
+                // Prepare the request for the interactor
+                let request = self.request_for_parallel_interactor(
+                    factor_sources
+                        .into_iter()
+                        .map(|f| f.factor_source_id())
+                        .collect(),
+                );
+                let response = interactor.sign(request).await?;
+                self.process_batch_response(response);
+            }
+
+            // Serial Interactor: One Factor Sources at a time
+            // After each factor source we pass the result to the collector
+            // updating its internal state so that we state about being able
+            // to skip the next factor source or not.
+            SigningInteractor::Serial(interactor) => {
+                for factor_source in factor_sources {
+                    // Prepare the request for the interactor
+                    let request =
+                        self.request_for_serial_interactor(&factor_source.factor_source_id());
+
+                    // Produce the results from the interactor
+                    let response = interactor.sign(request).await?;
+
+                    // Report the results back to the collector
+                    self.process_batch_response(response);
+
+                    if !self.continue_if_necessary()? {
+                        break;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn sign_with_factors_of_kind(
+        &self,
+        factor_sources_of_kind: FactorSourcesOfKind,
+    ) -> Result<()> {
+        let result = self.use_factor_sources(&factor_sources_of_kind).await;
+
+        match result {
+            Ok(_) => {}
+            Err(_) => self.process_batch_response(SignWithFactorSourceOrSourcesOutcome::Skipped {
```

Comment:
Is it ok to consider all errors as the factor source being skipped?

Reply:
Yes, because any Failure-To-Sign-With-FactorSource-X has the same effect on the process/final outcome as Skipped-FactorSource-X, in the sense that: no signatures from X are included. But it might be ok, imagine that FactorSource `X` was a an override factor next to `Y`, as long as we already have signed with `Y` or _will sign_ with `Y` the outcome of the collector can still be succesful.

Keep in mind that *retry* is out of scope for the collector. Any error reported back by the interactor (returned from the async methods in the interactor) means that Hosts have already prompted user for retry and user has skipped retry, or host has otherwise failed to retry.

Reply:
yes? sorry, what is your desired course of action? :)

Reply:
I guess it is ok as it is, I am not entirely sure that any error from the host should just be considered as skip. 
Can there be a potential conflict between the Wallet presenting some error alert to the user and then the collector requesting to present the next signature request? For example if a seed phrase is missing, when signing we do show an error alert, then I guess we should send the response back to Sargon only after user did dismiss that alert, or maybe we should not show that entirely. I guess we will see when integrating, but probably where I am getting at maybe the interactor should not return a Result, it either returns signed or skipped.

Reply:
@GhenadieVP @matiasbzurovski 

Fixed in https://github.com/radixdlt/sargon-mfa/pull/4 

new API:

```rust
pub trait SignWithFactorSerialInteractor {
    async fn sign(&self, request: SerialBatchSigningRequest) -> SignWithFactorsOutcome;
}
```

where:

```rust
pub enum SignWithFactorsOutcome {
    Signed {
        produced_signatures: BatchSigningResponse,
    },
    Neglected(NeglectedFactors), // (Reason, IndexSet<FactorSourceID>), where Reason is `Skipped | Failure`
}
```